src/lib/draw-and-pick.js

```diff
@@ -212,6 +263,93 @@ function pickFromBuffer(gl, {
   const width = Math.min(pickingFBO.width, deviceX + deviceRadius) - x + 1;
   const height = Math.min(pickingFBO.height, deviceY + deviceRadius) - y + 1;
 
+  const pickedColors = getPickedColors(gl, {layers, pickingFBO, deviceRect: {x, y, width, height}});
+
+  // Traverse all pixels in picking results and find the one closest to the supplied
+  // [deviceX, deviceY]
+  let minSquareDistanceToCenter = deviceRadius * deviceRadius;
+  let closestResultToCenter = null;
+  let i = 0;
+
+  for (let row = 0; row < height; row++) {
+    for (let col = 0; col < width; col++) {
+      // Decode picked layer from color
+      const pickedLayerIndex = pickedColors[i + 3] - 1;
+
+      if (pickedLayerIndex >= 0) {
+        const dx = col + x - deviceX;
+        const dy = row + y - deviceY;
+        const d2 = dx * dx + dy * dy;
+
+        if (d2 <= minSquareDistanceToCenter) {
+          minSquareDistanceToCenter = d2;
+
+          // Decode picked object index from color
+          const pickedColor = pickedColors.slice(i, i + 4);
+          const pickedLayer = layers[pickedLayerIndex];
+          const pickedObjectIndex = pickedLayer.decodePickingColor(pickedColor);
+          closestResultToCenter = {pickedColor, pickedLayer, pickedObjectIndex};
+        }
+      }
+      i += 4;
+    }
+  }
+
+  return closestResultToCenter || {
+    pickedColor: EMPTY_PIXEL,
+    pickedLayer: null,
+    pickedObjectIndex: -1
+  };
+}
+
+/**
+ * Query within a specified rectangle
+ * Returns array of unique objects in shape `{x, y, pickedColor, pickedLayer, pickedObjectIndex}`
+ */
+function queryFromBuffer(gl, {
+  layers,
+  pickingFBO,
+  deviceRect: {x, y, width, height}
+}) {
+  const pickedColors = getPickedColors(gl, {layers, pickingFBO, deviceRect: {x, y, width, height}});
+  const uniqueColors = new Map();
+
+  // Traverse all pixels in picking results and get unique colors
+  let i = 0;
+
+  for (let row = 0; row < height; row++) {
+    for (let col = 0; col < width; col++) {
+      // Decode picked layer from color
+      const pickedLayerIndex = pickedColors[i + 3] - 1;
+
+      if (pickedLayerIndex >= 0) {
+        const pickedColor = pickedColors.slice(i, i + 4);
+        const colorKey = pickedColor.join(',');
+        if (!uniqueColors.has(colorKey)) {
+          const pickedLayer = layers[pickedLayerIndex];
+          uniqueColors.set(colorKey, {
+            x: col + x,
+            y: row + y,
+            pickedColor,
+            pickedLayer,
+            pickedObjectIndex: pickedLayer.decodePickingColor(pickedColor)
+          });
+        }
+      }
+      i += 4;
+    }
+  }
+
+  return Array.from(uniqueColors.values());
+}
+/* eslint-enable max-depth, max-statements */
+
+// Returns an Uint8ClampedArray of picked pixels
+function getPickedColors(gl, {
+  layers,
+  pickingFBO,
+  deviceRect: {x, y, width, height}
+}) {
   // TODO - just return glContextWithState once luma updates
```

Comment:
I've worked around this TODO before as well -- @ibgreen , any context here? How low does this fruit hang?